test/extended/prometheus/prometheus.go

```diff
@@ -110,6 +110,25 @@ var _ = g.Describe("[sig-instrumentation][Late] Alerts", func() {
 		e2e.Logf("Total number of series sent via telemetry is below the limit")
 	})
 
+	g.It("shouldn't exceed the 200k number of series in the head block", func() {
+		oc.SetupProject()
+		ns := oc.Namespace()
+
+		execPod := exutil.CreateCentosExecPodOrFail(oc.AdminKubeClient(), ns, "execpod", nil)
+		defer func() {
+			oc.AdminKubeClient().CoreV1().Pods(ns).Delete(context.Background(), execPod.Name, *metav1.NewDeleteOptions(1))
+		}()
+
+		tests := map[string]bool{
+			// This is a rough current estimate of us to detect if we go over
+			// a given limit to possibly detect higher cardinality metrics.
+			`max_over_time(prometheus_tsdb_head_series[2h]) >= 200000`: false,
```

Comment:
This will cause problems when debugging by build watcher and other people without much knowledge into prometheus internals. Problems will be caused by lack of visibility into which job is causing cardinality explosion.

What about changing it to:
```
max_over_time(scrape_samples_scraped{job!~"kube-state-metrics|apiserver"}[2h]) >= 100000 OR max_over_time(prometheus_tsdb_head_series[2h]) >= 200000
```

This way test would catch if one particular job is causing cardinality explosion (most likely) and if there is no problem with the cumulative number of series. It also shows more metadata in case of test failure.

Reply:
Yes sounds good, but It only shows metadata for the failure if the job has more samples, so would solve it for that case, but if its multiple sources and only prometheus_tsdb_head_series is over that number it won't. What about if we also add a comment here how people can query, something like this above the test:

```
// In case of failures perform the following query in the Prometheus to see which are are the top
// 10 most samples scrapped per job:
// topk(10,max_over_time(scrape_samples_scraped{job!~"kube-state-metrics|apiserver|kubelet"}[2h]))
```

Note that
> max_over_time(scrape_samples_scraped{job!~"kube-state-metrics|apiserver"}[2h]) > 10000

is already failing :D for kubelet so will add that as well.

Reply:
> but It only shows metadata for the failure if the job has more samples

We won't be able to point directly why `prometheus_tsdb_head_series` is so high, but I think we can limit number of reports/BZs with additional check which shows directly what's the issue.

> What about if we also add a comment here how people can query

Good idea on a "test failure runbook"! However, code comments are not displayed in test failures and not many people look for failing test code when they see that test failing.

Reply:
I think we are really missing some metrics of series per scrape then right? :thinking: Metrics around samples will be most likely not helpful here.